[pytorch] Make process_group_agent catch some receive-side exceptions.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31299 [pytorch] Make process_group_agent catch some receive-side exceptions.**

Add try/catch around deserialization, and around the request callback, so
that these are converted to exceptions for process_group.

Differential Revision: [D19040826](https://our.internmc.facebook.com/intern/diff/D19040826/)

Commit:
[pytorch] Make process_group_agent catch some receive-side exceptions.

Add try/catch around deserialization, and around the request callback, so
that these are converted to exceptions for process_group.

Differential Revision: [D19040826](https://our.internmc.facebook.com/intern/diff/D19040826/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 19c8fab:
* **1/1** broken upstream at merge base ffe0c1a (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=ffe0c1ae4d8de75bc4d6cf5c43123df1e538d8a4&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=19c8fabf0549b2d8c49274678ffde45a61a6654a).

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [caffe2_onnx_py3_6_clang7_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=74518335)



---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "[pytorch] Make process_group_agent catch some receive-side exceptions."


Add try/catch around deserialization, and around the request callback, so
that these are converted to exceptions for process_group.

Differential Revision: [D19040826](https://our.internmc.facebook.com/intern/diff/D19040826/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -373,14 +373,33 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+              work.id_));
+        } catch (std::exception& e) {
+          LOG(ERROR) << "Could not deserialize " << payload.numel()
+                     << " byte payload from " << work.from_.id_;
```
Is it worth it to try sending a response back to `work.from_id` indicating that it's message couldn't be deserialized?

jjlilley replied:
Good idea, seems like it would be more reasonable to send a relevant exception back.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -414,6 +435,8 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
                 });
           }
         } else if (message.isResponse()) {
+          // Response handling
```
I think there are also some exceptions that can happen here (in response handling).  For example with the changes I added in https://github.com/pytorch/pytorch/pull/31381, a runtime error was being raised in `future->markCompleted` and the program would just hang, which made it a lot harder to debug. Catching the exception there could improve the debugging experience.

jjlilley replied:
Good catch, I'll cover the markCompleted()/setError() section below where it's possible that the  callback might accidentally throw.

Commit:
Update on "[pytorch] Make process_group_agent catch some receive-side exceptions."


Add try/catch around deserialization, and around the request callback, so
that these are converted to exceptions for process_group.

Differential Revision: [D19040826](https://our.internmc.facebook.com/intern/diff/D19040826/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -373,14 +373,45 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
   threadPool_.run(std::bind(
       [&](RecvWork& work) {
         torch::Tensor& payload = work.payload_;
```
Can we move this entire code block to a function `handleRecv(RecvWork)` and then do something like this in this function:

```
try {
  handleRecv(work);
} catch (std::exception& e) {
  // Based on whether work is a request/response we handle the exception appropriately.
}

I feel this would be much cleaner than having individual try-catch blocks spread throughout this function.

jjlilley replied:
From a glance, one might think so. But practically speaking, there are roughly 3 specific classes of errors that I believe we want to handle here, in different ways:

1.  Deserialization errors - this is what initially motivated the change. In this case, for server-side we want to return an exception response to the client, or on client-side, just to fulfill the future with an exception.

2. server-side processing errors- suppose the request_callback_impl throws an exception. Thrift handles this ok currently, but process group should catch and channel this back into an exception for the client to see (like Thrift basically does).

3. errors fulfilling the future- this one shouldn't really happen, but Rohan pointed as a possibility, and it's probably best to just log when the added callback itself throws so we can address.

jjlilley replied:
Thanks for comments, I'll see if I can add a few test cases somewhere.
I'm debating where to put? Is it acceptable to make fb-only tests here (which would be somewhat simpler), or should they be in OSS?

pritamdamania87 replied:
I feel a large try-catch block is safer since it ensures we don't miss out on any cases as Rohan pointed out here https://github.com/pytorch/pytorch/pull/31299#discussion_r359042760. Also, as we modify the logic here in the future its safer to have one try-catch block at the top level which handles all errors. Otherwise, it's easy for developers in the future to add some functionality but they forget to handle exceptions. Also, the single try-catch block won't just catch and log, it should send a response back to the client if its a request and mark an appropriate error on the future if it is a response.

I think the tests should be in OSS since this is OSS code we're modifying.

jjlilley replied:
Let me think about how to add coverage here - it's fair to ask for it, though OSS cpp code coverage is quite poor vs. fb-only.

In any case, having an overall try/catch block for 3 (and maybe 2), might well be reasonable. For the  1. case above, I think there's a good argument for more specialized handling. In any case, let's see when we can get this up to parity with Thrift.

rohan-varma replied:
@jjlilley For now, we (unfortunately) don't have cpp tests for process group agent, there's an issue to add them: https://github.com/pytorch/pytorch/issues/28309 but we probably won't get to this for some time. For now, I'm wondering if you can add them in python similar to what I did in https://github.com/pytorch/pytorch/pull/31023, but simulating an exception on the recv side might be trickier

jjlilley replied:
Yeah, thanks, I know process_group_agent has zero cpp test coverage.
I've seen the pattern - people are all quietly hoping somebody else adds it :)

fwiw, I did write up some c++ ProcessGroupAgent benchmarking code in fb-only the other month, when I was trying to optimize rpc perf:
  caffe2/torch/fb/distributed/thriftRpcBackend/test/ProcessGroupAgentBench.cpp

I might try to massage that into something useful for test coverage in OSS.

pritamdamania87 reviewed (changes requested):
Can we add some unit tests for this to ensure exceptions are handled appropriately?

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -373,14 +373,45 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+          LOG(ERROR) << "Could not deserialize " << payload.numel()
+                     << " byte payload from " << work.from_.id_;
+          Message emptyBaseMsg({}, {}, work.type_, work.id_);
+          Message exceptionResponse = createExceptionResponse(
```
Can user distinguish whether failure occurs on caller or callee based on this error message? If not, shall we also add `work.from_.id_` here?

mrshenli reviewed (commented):
Thanks for working on this! This would help improve debugging experiences especially when making changes in modules that RPC depends on. cc @zdevito 

Regarding the C++ tests for process group agent, I can add that if no one has picked it up. @jjlilley have you started working on this?

rohan-varma commented:
> Regarding the C++ tests for process group agent, I can add that if no one has picked it up. @jjlilley have you started working on this?

There are C++ tests now in the internal codebase, and I am working on migrating them to OSS, see https://github.com/pytorch/pytorch/pull/32152

jjlilley commented:
> Thanks for working on this! This would help improve debugging experiences especially when making changes in modules that RPC depends on. cc @zdevito
> 
> Regarding the C++ tests for process group agent, I can add that if no one has picked it up. @jjlilley have you started working on this?

@mrshenli - I did write an initial process group agent test.

I really wanted to check it into caffe2/test/cpp/rpc somewhere, but had brutal fight with CMake - simply could find no way to get it to pass the linker, even with TORCH_API thrown around/etc. Burned waaay more time at this than it was worth and checked into fb-only.

It's currently checked in caffe2/torch/fb/distributed/thriftRpcBackend/test/ProcessGroupAgentTest.cpp

That said, if you think you can get it to link, it would be awesome to move it to the open source version.

rohan-varma commented:
@jjlilley I was wondering what's the status on this, is it going to be land-able soon? Would really help in closing https://github.com/pytorch/pytorch/issues/30819

jjlilley commented:
> @jjlilley I was wondering what's the status on this, is it going to be land-able soon? Would really help in closing #30819

@rohan-varma Based on earlier feedback, it's not entirely clear what it would take for this change to be land-able, hence my putting this on the back-burner. I'm happy to iterate here if agreement is possible.

- I did land an fb-side c++ unittest for process_group (see caffe2/torch/fb/distributed/thriftRpcBackend/test/ProcessGroupAgentTest.cpp), and some of the test-cases make me feel a bit better about the main exception cases that I was thinking about.

- The main problematic cases I expect here are the deserialization failures; but it's not trivially possible to add tests for these (at least without adding indirection that it wasn't clear was wanted). And it seemed that without clear tests, that the review was going to be blocked. Am I misunderstanding this?

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -443,11 +478,15 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+              fm->markCompleted(std::move(message));
+            }
+          } catch (std::exception& e) {
+            LOG(ERROR) << "Uncaught exception when marking rpc complete";
```
nit: log what the exception here is? `e.what()`

rohan-varma commented:
@pritamdamania87 @mrshenli Considering @jjlilley's comments, do you have any suggestions on making progress to getting this PR into a landeable state? It looks like this is best tested from the C++ side as the modifications are related to internal `ProcessGroupAgent` implementation details which aren't surfaced to the python layer. 

Given that it is challenging to port the C++ PG agent tests to OSS (@jjlilley tried and I tried in https://github.com/pytorch/pytorch/pull/32263, although it is tough to debug the linker errors), is it okay to keep tests for this in internal code? Otherwise, is there a better way to make progress on this PR?
